### PR TITLE
feat(message): add message component

### PR DESCRIPTION
**Pull request by 0x74616e67**
## Checklist

- [ ] Fix linting errors
- [ ] Tests have been added / updated (or snapshots)

### CODE REVIEW

File: components/use-messages/use-message.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+export interface Message {
+  icon?: React.ReactNode
+  text?: string
+  color?: MessageColors
+  delay?: number
+  closeable?: boolean
+  shadow?: boolean
+  className?: string
+  onClose?: Function
```
Comment (yqrashawn) [inline components/use-messages/use-message.tsx]: better use `() => void`

File: components/use-messages/message-item.tsx [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+          .message:first-child {
+            margin-top: 0;
+          }
+          .message.visible {
+            transform: translate(0, 0);
+            opacity: 1;
+            z-index: 2;
+          }
+          .message.hide {
+            margin-top: -3.5714rem;
```
Comment (yqrashawn) [inline components/use-messages/message-item.tsx:104]: Why not use the layout.gap variable here

File: components/use-messages/message-item.tsx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      const timer = setTimeout(() => {
+        setVisible(true)
+        clearTimeout(timer)
+      }, 10)
+      return () => clearTimeout(timer)
+    }, [])
+    useEffect(() => {
+      let timer: any = null
+      if (hover) {
+        clearTimeout(timer)
```
Comment (yqrashawn) [inline components/use-messages/message-item.tsx]: timer here will always be `null`

Comment (yqrashawn) [reply components/use-messages/message-item.tsx]: maybe use ref to store timer, there's similar code in auto-complete

File: components/use-messages/message-item.tsx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    const [hover, setHover] = useState<boolean>(false)
+    useEffect(() => {
+      const timer = setTimeout(() => {
+        setVisible(true)
+        clearTimeout(timer)
+      }, 10)
+      return () => clearTimeout(timer)
+    }, [])
+    useEffect(() => {
+      let timer: any = null
```
Comment (yqrashawn) [inline components/use-messages/message-item.tsx]: Don't use `any` here.

File: components/use-messages/message-item.tsx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            clearTimeout(timer)
+          }, delay)
+        }
+      }
+      return () => {
+        delay && clearTimeout(timer)
+      }
+    }, [delay, hover])
+    useEffect(() => {
+      let timer: any = null
```
Comment (yqrashawn) [inline components/use-messages/message-item.tsx]: Don't use `any` here.

File: components/use-messages/message-item.tsx [hunk 6]
```diff
@@ -0,0 +1,148 @@
+import React, { useEffect, useMemo, useState } from 'react'
+import useTheme from '../styles/use-theme'
+import { Message } from './use-message'
+import X from '@zeit-ui/react-icons/x'
+import { getStyles } from './styles'
+
+export type MessageItemProps = Message & {
+  id?: string
+  destroy?: Function
```
Comment (yqrashawn) [inline components/use-messages/message-item.tsx]: Use `(id: string) => void` here.

File: components/use-messages/use-message.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+}
+const defaultMessage = {
+  delay: 2000,
+  color: 'default' as MessageColors,
+  closeable: false,
+  shadow: true,
+  onClose: () => {},
```
Comment (yqrashawn) [inline components/use-messages/use-message.tsx]: Why use an empty function here?

**Review by yqrashawn — CHANGES_REQUESTED** [review]

File: components/use-messages/message-item.tsx (hunk 2 above)
Comment (0x74616e67) [reply components/use-messages/message-item.tsx:104]: It is message component height value, not gap value, should not affect with gap config.

File: components/use-messages/use-message.tsx (hunk 1 above)
Comment (0x74616e67) [reply components/use-messages/use-message.tsx]: done

File: components/use-messages/message-item.tsx (hunk 6 above)
Comment (0x74616e67) [reply components/use-messages/message-item.tsx]: done

File: components/use-messages/message-item.tsx (hunk 3 above)
Comment (0x74616e67) [reply components/use-messages/message-item.tsx]: done

File: components/use-messages/message-item.tsx (hunk 4 above)
Comment (0x74616e67) [reply components/use-messages/message-item.tsx]: done

File: components/use-messages/message-item.tsx (hunk 5 above)
Comment (0x74616e67) [reply components/use-messages/message-item.tsx]: done

File: components/use-messages/use-message.tsx (hunk 7 above)
Comment (0x74616e67) [reply components/use-messages/use-message.tsx]: done

**Review by yqrashawn — APPROVED** [review]

**Review by yqrashawn — APPROVED** [review]

**Comment by AngeliaQ** [comment]
Task linked: [CU-3tvnwx Message component](https://app.clickup.com/t/3tvnwx)

**Review by AngeliaQ — APPROVED** [review]

**Pull request merged**